modules/subchannel/src/problems/SubChannel1PhaseProblem.C

```diff
@@ -1906,215 +1953,112 @@ SubChannel1PhaseProblem::computeAddedHeatDuct(unsigned int i_ch, unsigned int iz
-  }
-  else
+    LibmeshPetscCall(VecDuplicate(src, dst));
+    LibmeshPetscCall(VecCopy(src, *dst));
+  };
+
+  // per-field relax+blend helper with explicit return type.
+  auto RelaxField =
+      [&](Mat A_ff, Vec rhs_f, Vec like_vec, Vec work, PetscScalar alpha, auto && populate)
+      -> PetscErrorCode
   {
-    unsigned int last_node = (iblock + 1) * _block_size;
-    unsigned int first_node = iblock * _block_size + 1;
-    LibmeshPetscCall(populateVectorFromHandle<SolutionHandle>(
-        _prod, *_mdot_soln, first_node, last_node, _n_channels));
-    Vec ls;
-    LibmeshPetscCall(VecDuplicate(_amc_sys_mdot_rhs, &ls));
-    LibmeshPetscCall(MatMult(_amc_sys_mdot_mat, _prod, ls));
-    LibmeshPetscCall(VecAXPY(ls, -1.0, _amc_sys_mdot_rhs));
-    LibmeshPetscCall(VecAXPY(vec_array[field_num], -1.0, ls));
-    LibmeshPetscCall(VecDestroy(&ls));
-  }
+    Vec d = nullptr;
+    PetscCall(VecDuplicate(like_vec, &d));
 
-  if (_verbose_subchannel)
-    _console << "Lin mom OK." << std::endl;
+    // 1) A_ff: diag <- diag / alpha
+    PetscCall(MatGetDiagonal(A_ff, d));
+    PetscCall(VecScale(d, 1.0 / alpha));
+    PetscCall(MatDiagonalSet(A_ff, d, INSERT_VALUES));
 
-  // Cross momentum conservation
-  field_num = 2;
-  mat_array[Q * field_num + 0] = NULL;
-  if (_pressure_cross_momentum_tight_coupling)
-  {
-    LibmeshPetscCall(
-        MatDuplicate(_cmc_pressure_force_mat, MAT_COPY_VALUES, &mat_array[Q * field_num + 1]));
-    LibmeshPetscCall(MatAssemblyBegin(mat_array[Q * field_num + 1], MAT_FINAL_ASSEMBLY));
-    LibmeshPetscCall(MatAssemblyEnd(mat_array[Q * field_num + 1], MAT_FINAL_ASSEMBLY));
-  }
-  else
-  {
-    mat_array[Q * field_num + 1] = NULL;
-  }
+    // 2) work <- x_old (caller-provided populator)
+    PetscCall(populate(work));
 
-  LibmeshPetscCall(MatDuplicate(_cmc_sys_Wij_mat, MAT_COPY_VALUES, &mat_array[Q * field_num + 2]));
-  LibmeshPetscCall(MatAssemblyBegin(mat_array[Q * field_num + 2], MAT_FINAL_ASSEMBLY));
-  LibmeshPetscCall(MatAssemblyEnd(mat_array[Q * field_num + 2], MAT_FINAL_ASSEMBLY));
-  if (_monolithic_thermal_bool)
-  {
-    mat_array[Q * field_num + 3] = NULL;
-  }
+    // 3) rhs_f += (1 - alpha) * (diag .* work)
+    PetscCall(VecScale(d, (1.0 - alpha)));
+    PetscCall(VecPointwiseMult(work, work, d));
+    PetscCall(VecAXPY(rhs_f, 1.0, work));
 
-  LibmeshPetscCall(VecDuplicate(_cmc_sys_Wij_rhs, &vec_array[field_num]));
-  LibmeshPetscCall(VecCopy(_cmc_sys_Wij_rhs, vec_array[field_num]));
-  if (_pressure_cross_momentum_tight_coupling)
-  {
-    LibmeshPetscCall(VecAXPY(vec_array[field_num], 1.0, _cmc_pressure_force_rhs));
-  }
-  else
-  {
-    Vec sol_holder_P;
-    LibmeshPetscCall(createPetscVector(sol_holder_P, _block_size * _n_gaps));
-    LibmeshPetscCall(populateVectorFromHandle<SolutionHandle>(
-        _prodp, *_P_soln, iblock * _block_size, (iblock + 1) * _block_size - 1, _n_channels));
+    PetscCall(VecDestroy(&d));
```

Comment:
```suggestion
    LibmeshPetscCall(VecDestroy(&d));
```
same for all the others please